[DYOD18/19] Unique Constraints

nilsthamm opened the pull request:
Current status of the Primary Key Groups work.

The basic unique and concatenated unique constraint functionality is added. Unique constraints can be added on a table and their definition consists of the respecting column IDs and if the constraint is a primary key.

`void Table::add_unique_constraint(const std::vector<ColumnID>& column_ids, bool primary)`

As discussed in the last IMDB roundtable meeting. The checker is called, only if a constraint is defined for a coulmn, during the execution of the insert operator and during the commit process in the transaction context.
NOTE: Due to the bug described in #1393 and the necessary refactoring of the commit process, a transition back to the state active is currently used as soon as a violated constraint is detected during the commit process (see `TransactionContext::commit_async`)

Currently the uniqueness is tested on a brute force way, by creating a set out of the whole column. The next steps therefore will be to refactor and optimize the unique_checker.cpp file, so this file is kind of work in progress.

mrks commented on src/lib/storage/constraints/table_constraint_definition.hpp:
```diff
@@ -0,0 +1,14 @@
+
+struct TableConstraintDefinition final {
+  std::vector<ColumnID> columns;
+  bool is_primary_key;
```
Do we check somewhere that a table has only a single primary key?

mrks replied:
Whoops, that was meant to be a pending comment ;)

mrks commented:
@d-justen @engelfa @Miroka96 - Please remember to review this PR. If possible, I'd like to get this of the plate before the next wave of PRs comes in.

julianmenzler commented on src/benchmark/operators/insert_constraint_benchmark.cpp:
```diff
@@ -0,0 +1,117 @@
+  }
+}
+
+static void PreInsertRanges(benchmark::internal::Benchmark* b) {
```
Simplify naming here and use a function name similar to the one declared above. E.g.:
```suggestion
static void InsertRangesFilledTable(benchmark::internal::Benchmark* b) {
```

mrks replied:
No idea what this is supposed to do.

julianmenzler commented on src/test/storage/constraints_test.cpp:
```diff
@@ -0,0 +1,376 @@
+
+    nullable_column_definitions.emplace_back("column0", DataType::Int, true);
+    nullable_column_definitions.emplace_back("column1", DataType::Int, false);
+    nullable_column_definitions.emplace_back("column2", DataType::Int, false);
```
I'm a little irritated by the naming. Both column definition structs `column_definitions`& `nullable_column_definitions` define two non-nullable and a one nullable column. 
The naming implies more fundamental differences. Or am I misunderstanding something?

Bensk1 commented:
> @d-justen @engelfa @Miroka96 - Please remember to review this PR. If possible, I'd like to get this of the plate before the next wave of PRs comes in.

@Miroka96 seems to work on the between/like replacement topic

mrks commented:
Mmmh, then where did I get that user name from?

nilsthamm commented on src/lib/storage/constraints/unique_checker.hpp:
```diff
@@ -0,0 +1,28 @@
+namespace opossum {
+
+bool constraint_valid_for(const Table& table, const TableConstraintDefinition& constraint,
+                          const CommitID& snapshot_commit_id, const TransactionID& our_tid);
```
Good point we changed that in our current version of the checker

nilsthamm commented on src/lib/operators/abstract_read_write_operator.hpp:
```diff
@@ -92,6 +96,8 @@ class AbstractReadWriteOperator : public AbstractOperator {
    */
   void _mark_as_failed();
 
+  const std::string _target_table_name;
```
During the `_commit_async` in _transaction_context.cpp_ we need to access the `table_name()` function from the abstract read write operator interface. Additionally, it turned out all of the read write were already storing this name as a member variable but under different names.

mrks replied:
As discussed: dynamic_pointer_cast

nilsthamm commented on src/benchmark/operators/insert_constraint_benchmark.cpp:
```diff
@@ -0,0 +1,117 @@
+  }
+}
+
+static void InsertRanges(benchmark::internal::Benchmark* b) {
```
Changed that (and the other suggestion) in the current version of our benchmark 👍

nilsthamm commented:
@d-justen @julianmenzler @engelfa I think we addressed all comments or applied them directly. Also, all merge conflicts are resolved. Therefore, we removed the checking based on the table name and all related functions we added. So please have a second look!

julianmenzler commented on src/lib/concurrency/transaction_context.cpp:
```diff
@@ -76,6 +78,18 @@ bool TransactionContext::commit_async(const std::function<void(TransactionID)>&
+    const auto& type = op->type();
+    // TOOD(all): Remove as soon as the transaction context phase model got refactored
+    if ((type == OperatorType::Update || type == OperatorType::Insert) &&
+        !all_constraints_valid_for(op->table_name(), _commit_context->commit_id(), _transaction_id)) {
```
You're right, got it!

>In our current version, we are skipping all immutable chunks during this check to only look at those who might have conflicts.

Great idea! This should reduce the impact on performance drastically. 👍

mrks commented on src/test/storage/constraints_test.cpp:
```diff
@@ -0,0 +1,409 @@
+
+    table_temp->append({1, 1, 3});
+    table_temp->append({2, 1, 2});
+    table_temp->append({3, 2, 0});
```
Having read further, I wonder where the tables and ops defined here are used. I'll review the tests once this is cleaned up.

mrks commented on src/benchmark/operators/insert_constraint_benchmark.cpp:
```diff
@@ -0,0 +1,117 @@
+
+namespace opossum {
+
+auto reset_table(bool use_constraints, int values_to_insert, int num_rows = 0) {
```
We don't use `auto` for return types (except for lambdas and where unavoidable). I shouldn't have to read the method to understand its signature.

mrks replied:
Isn't this simply a setup method that should be part of a fixture?

mrks commented on src/benchmark/operators/insert_constraint_benchmark.cpp:
```diff
@@ -0,0 +1,117 @@
+  manager.add_table("pre_insert_table_temp", pre_insert_table_temp);
+
+  for (; row_preinserted < num_rows; row_preinserted++) {
+    pre_insert_table_temp->append({row_preinserted, row_preinserted * 2});
```
Do I understand the setup correctly that you do not directly append to the table because you want the MVCC data to be set?

mrks commented on src/benchmark/operators/insert_constraint_benchmark.cpp:
```diff
@@ -0,0 +1,117 @@
+  pre_insert->execute();
+  pre_insert_context->commit();
+
+  // Create insert operators depending on values to insert operator
```
This is a completely different concern. How does this relate to "resetting the table"?

mrks commented on src/benchmark/operators/insert_constraint_benchmark.cpp:
```diff
@@ -0,0 +1,117 @@
+  return table_inserts;
+}
+
+BENCHMARK_DEFINE_F(MicroBenchmarkBasicFixture, BM_InsertEmptyTableWithConstraint)(benchmark::State& state) {
```
A quick comment on what these benchmarks measure would be great.

mrks replied:
Isn't this the same benchmark as the one below, just with the number of pre-inserted values being zero?

mrks commented on src/lib/concurrency/transaction_context.cpp:
```diff
@@ -76,6 +78,18 @@ bool TransactionContext::commit_async(const std::function<void(TransactionID)>&
 
   if (!success) return false;
 
+  // Check all _rw_operators potential violations of unique constraints.
```
**for** potential violations?

mrks commented on src/lib/concurrency/transaction_context.cpp:
```diff
@@ -76,6 +78,18 @@ bool TransactionContext::commit_async(const std::function<void(TransactionID)>&
+  // If the constraint check fails, set the commit as failed.
+  for (const auto& op : _rw_operators) {
+    const auto& type = op->type();
+    // TOOD(all): Remove as soon as the transaction context phase model got refactored
```
I am not sure that an outsider would know how to address this todo.

mrks commented on src/lib/concurrency/transaction_manager.hpp:
```diff
@@ -61,7 +61,8 @@ class TransactionManager : public Singleton<TransactionManager> {
   // deleted in the same transaction (which has not yet committed)
   static constexpr auto INVALID_TRANSACTION_ID = TransactionID{0};
-  static constexpr auto INITIAL_TRANSACTION_ID = TransactionID{1};
+  static constexpr auto UNUSED_TRANSACTION_ID = TransactionID{1};
```
Add a comment on what this is good for?

mrks commented on src/lib/operators/abstract_read_write_operator.cpp:
```diff
@@ -63,4 +66,9 @@ void AbstractReadWriteOperator::_mark_as_failed() {
 }
 
+const std::string AbstractReadWriteOperator::table_name() {
+  Assert(type() == OperatorType::Insert || type() == OperatorType::Update, "Delte operators don't store table names");
```
typo

mrks commented on src/lib/operators/insert.cpp:
```diff
@@ -214,6 +215,13 @@ std::shared_ptr<const Table> Insert::_on_execute(std::shared_ptr<TransactionCont
   }
 
+  if (transaction_context_is_set()) {
+    if (!all_constraints_valid_for(_target_table_name, transaction_context()->snapshot_commit_id(),
```
"all" seems redundant here. Maybe `constraints_satisfied(...)`?

mrks commented on src/lib/storage/constraints/unique_checker.cpp:
```diff
@@ -0,0 +1,63 @@
+
+bool constraint_valid_for(const Table& table, const TableConstraintDefinition& constraint,
+                          const CommitID& snapshot_commit_id, const TransactionID& our_tid) {
+  std::set<boost::container::small_vector<AllTypeVariant, 3>> unique_values;
```
Was `std::set` the fastest here? We might have had this discussion before, but did you try it with `unordered_set` and `(vector.sort().unique()).size() == vector.size()`?

m0r13 replied:
We're going to use a sorted vector in the optimized next version.

mrks commented on src/lib/storage/constraints/unique_checker.cpp:
```diff
@@ -0,0 +1,63 @@
+      const auto begin_cid = mvcc_data->begin_cids[chunk_offset];
+      const auto end_cid = mvcc_data->end_cids[chunk_offset];
+
+      auto row = boost::container::small_vector<AllTypeVariant, 3>();
```
add a `using` for the type?

mrks commented on src/lib/storage/constraints/unique_checker.cpp:
```diff
@@ -0,0 +1,63 @@
+      if (Validate::is_row_visible(our_tid, snapshot_commit_id, row_tid, begin_cid, end_cid)) {
+        for (const auto& column_id : constraint.columns) {
+          const auto& segment = segments[column_id];
+          const auto& value = segment->operator[](chunk_offset);
```
I use `(*segment)[n]`, but we don't really have a rule for it.

m0r13 replied:
Just doing `(*segment)[n]` now.

mrks commented on src/lib/storage/constraints/unique_checker.cpp:
```diff
@@ -0,0 +1,63 @@
+      const auto end_cid = mvcc_data->end_cids[chunk_offset];
+
+      auto row = boost::container::small_vector<AllTypeVariant, 3>();
+      row.reserve(constraint.columns.size());
```
Writing to a pre-resized vector using `[]=` is generally also faster

m0r13 replied:
Going to set the size directly to the number of columns in the constraint, and writing to it with `[]=`.

@julianmenzler what you suggest is not possible as that `3` is a template argument, and can't be determined by the number of columns in the constraint.

mrks commented on src/lib/storage/constraints/unique_checker.cpp:
```diff
@@ -0,0 +1,63 @@
+  return true;
+}
+
+bool all_constraints_valid_for(std::shared_ptr<const Table> table, const CommitID& snapshot_commit_id,
```
pass by reference?

mrks commented on src/lib/storage/constraints/unique_checker.hpp:
```diff
@@ -0,0 +1,28 @@
+
+bool constraint_valid_for(const Table& table, const TableConstraintDefinition& constraint,
+                          const CommitID& snapshot_commit_id, const TransactionID& our_tid);
+bool all_constraints_valid_for(std::shared_ptr<const Table> table, const CommitID& snapshot_commit_id,
```
Consistency: Why do you sometimes pass around `shared_ptr<Table>`, and sometimes `Table&`?

m0r13 replied:
Originally we started with `shared_ptr<Table>`, but `constraint_valid_for` needs to be called from inside `table.add_unique_constraint` where we don't have a shared pointer. Should we use const reference everywhere?

m0r13 commented on src/lib/storage/constraints/unique_checker.cpp:
```diff
@@ -0,0 +1,63 @@
+
+bool constraint_valid_for(const Table& table, const TableConstraintDefinition& constraint,
+                          const CommitID& snapshot_commit_id, const TransactionID& our_tid) {
+  std::set<boost::container::small_vector<AllTypeVariant, 3>> unique_values;
```
Adding a comment about that. `small_vector` because of preallocation, yes. Three slots because we think that three is the maximum number of columns most unique constraints will use.

m0r13 commented on src/lib/concurrency/transaction_context.cpp:
```diff
@@ -76,6 +78,18 @@ bool TransactionContext::commit_async(const std::function<void(TransactionID)>&
+  for (const auto& op : _rw_operators) {
+    const auto& type = op->type();
+    // TOOD(all): Remove as soon as the transaction context phase model got refactored
+    if ((type == OperatorType::Update || type == OperatorType::Insert) &&
```
Yep, it's fixed by now.

m0r13 commented on src/lib/operators/update.cpp:
```diff
@@ -55,7 +54,7 @@ std::shared_ptr<const Table> Update::_on_execute(std::shared_ptr<TransactionCont
+  _insert = std::make_shared<Insert>(_target_table_name, _input_right);
   _insert->set_transaction_context(context);
   _insert->execute();
   // Insert cannot fail in the MVCC sense, no check necessary
```
Fixed.

m0r13 commented:
The feedback should be adressed now for everything except the tests and benchmarks. Note also that we changed the target branch to `constraints/master`, not the real Hyrise master. We would refactor the tests and benchmarks and bring them in properly with the next PR along with the various performance optimizations.

mrks commented on src/lib/storage/constraints/unique_checker.cpp:
```diff
@@ -0,0 +1,67 @@
+                          const CommitID& snapshot_commit_id, const TransactionID& our_tid) {
+  // we store the value tuples of unique columns in a set to check if the values are unique
+  // each tuple is a boost small_vector with three elements already preallocated on the stack
+  // three because we think that most constraints use a maximum of three columns
```
A little regard for punctuation and capitalization, please?

mrks commented on src/lib/storage/constraints/unique_checker.cpp:
```diff
@@ -0,0 +1,67 @@
+
+namespace opossum {
+
+using tuple_row = boost::container::small_vector<AllTypeVariant, 3>;
```
Surprised that clang-tidy lets you do this. Types are CamelCase.

mrks commented on src/benchmark/operators/insert_constraint_benchmark.cpp:
```diff
@@ -0,0 +1,117 @@
+  }
+}
+
+static void InsertRanges(benchmark::internal::Benchmark* b) {
```
Code style

mrks commented on src/lib/storage/table.hpp:
```diff
@@ -161,16 +162,24 @@ class Table : private Noncopyable {
   size_t estimate_memory_usage() const;
 
+  /**
+   * Add a unique constraint. Column IDs should be sorted.
```
But *you* are sorting them?!?

m0r13 replied:
Oh alright. Fixing that comment.

mrks commented on src/lib/storage/table.cpp:
```diff
@@ -212,4 +216,39 @@ size_t Table::estimate_memory_usage() const {
+
+  // Since we don't work with a transaction context here we need to make sure no other operation adds a value before
+  // we finished our check
+  auto append_lock = acquire_append_mutex();
```
Please add an explicit scope when you use an RAII-style lock

mrks commented on src/lib/storage/constraints/unique_checker.cpp:
```diff
@@ -0,0 +1,67 @@
+bool constraints_satisfied(const Table& table, const CommitID& snapshot_commit_id,
+                               const TransactionID& our_tid) {
+  for (const auto& constraint : table.get_unique_constraints()) {
+    if (!constraint_valid_for(table, constraint, snapshot_commit_id, our_tid)) {
```
Consistency: `/constraints?_satisfied/`

m0r13 replied:
`constraints_satisfied` checks if all constraints on a table are satisfied, `constraint_valid_for` checks if a single constraint will be still valid for some inserted values, if you were wondering about the plural. But perhaps the `constraint_valid_for` is not the best name to check a constraint after inserting values, we can think about some better names here.

mrks replied:
My comment was about "satisfied" vs "valid_for". Why not `constraint_satisfied` and `constraints_satisfied`?

m0r13 replied:
Makes sense, I renamed it.

nilsthamm commented:
@mrks we addressed everything, should we merge? 
@julianmenzler Do you have any more change requests? If not please dismiss your review or approve, otherwise merging is blocked 😉

mrzzzrm commented:
Can I block this until tonight? I'd at least like to give it a quick glance-over.

mrzzzrm commented:
I changed the base branch to `hyrise:master` because the diff was showing a lot of changes not made by this PR.

Also, there seem to be unadressed comments by @mrks (https://github.com/hyrise/hyrise/pull/1416/files#r253855165) - are those actually done? if so, it's best to resolve the conversation, ideally with a closing comment.

nilsthamm commented:
@mrzzzrm FYI: that this is not our final implementation. In our next PR we are going to refactor the tests and benchmarks (this includes @mrks comment) and also in the unique checker, we changed a lot. 
Regarding the new base. That is right, I merged the wrong branch. The idea was to not merge this not yet fully finished state of our implementation into the master. My suggestion would be to change the base back to the constraint master (but probably reset it first to the current master)

mrzzzrm commented:
So you actually wanted to merge this into `constraints/master`, not the actual `master`? Then I should probably keep my nose out of this and wait until a PR for the actual master is ready.

nilsthamm commented:
Yes, the idea was that the other group doesn't have to review the already reviewed changes again, when we add our new code (since this PR is already pretty old). If you want to see the full and final picture, you are right, you better look at the final PR from this branch into the master

mrzzzrm commented:
\* pulls his nose out of this PR and pretends he did not say anything \*
